Fix: added support for both parameters (async & async_) in the public API

unimariJo opened the pull request:
# Introduction
`async` parameter is parsed as the part of asyncio module, but also this parameter needed to make an asynchronous connection. 

# Solution
To fix the error, we should replace the name of parameter. The official psycopg2 uses `async_` parameter, so it will be better to us too to replace the parameter's name to `async_` to avoid confusion.

lopuhin commented:
@unimariJo async becomes a keyword in python 3.7, right? Also, isn't `async_` added as an extra alias in addition to `async` in psycopg2?

unimariJo commented:
@lopuhin Yes, async becomes a keyword in python 3.7 but if asyncio is imported, `async` becomes an expression (therefore we anyway need different keyword). And you are right that `async_` is an extra alias (I forgot about that) in psycopg2, so I can make it as the alias in my fork too. Should I do this?

lopuhin commented:
> Yes, async becomes a keyword in python 3.7 but if asyncio is imported, async becomes the keyword too. 

Oh, I didn't know that. Do you have an example that shows that? I tried the following in python 3.6 and it seems to work fine:

```
In [5]: import asyncio
In [6]: async = 1
In [7]: def foo(async):
   ...:     return async
   ...:
In [8]: foo(42)
Out[8]: 42
```

> And you are right that async_ is an extra alias (I forgot about that) in psycopg2, so I can make it as the alias in my fork too. Should I do this?

Yes, it would be ideal if you can make it support both ``async`` and ``async_`` in the public API (from a quick glance it looks like only ``connect`` function from ``psycopg2cffi/__init__.py`` is public among the changed ones). To support python 3.7, you can fetch ``async`` from kwargs.

unimariJo commented:
@lopuhin I made support for both parameters (`async` & `async_`). Yay! 
P.S I will look for examples.

lopuhin commented on psycopg2cffi/_impl/connection.py:
```diff
@@ -121,7 +121,13 @@ def __init__(self, dsn, async=False):
+        if 'async_' in kwargs:
+            self._async = kwargs.pop('async_')
+        else:
+            self._async = False
```
this will overwrite passed ``async`` value (if ``async`` is passed by ``async_`` isn't, ``self._async`` will always be False

lopuhin replied:
Will be also nice to check that kwargs are empty after we determine ``self._async`` value, else unsupported ``kwargs`` will be silently ignored instead of rising the error (e.g. ``Connection(do_what_i_want=True)`` will not raise an error after this change).

lopuhin commented on psycopg2cffi/__init__.py:
```diff
@@ -41,7 +41,7 @@ def _param_escape(s,
 
 def connect(dsn=None,
         database=None, user=None, password=None, host=None, port=None,
-        connection_factory=None, cursor_factory=None, async=False, **kwargs):
```
It will be better if ``async_`` is present as a keyword in the function signature: it's functionally equivalent but it's more friendly to IDEs and people reading the code.

unimariJo replied:
When I added `async_` as a keyword in the function, tests haven't been passed so it isn't functionally equivalent ([commit](https://github.com/chtd/psycopg2cffi/pull/92/commits/772fe9244c9006b4255374f7fa1dc5ab56dba127), [Travis](https://travis-ci.org/chtd/psycopg2cffi/builds/304747230?utm_source=github_status&utm_medium=notification))

lopuhin commented:
@unimariJo thanks, this looks good overall 👍
I left a couple of comments above.

unimariJo commented:
@lopuhin I fixed all remarks in yours reviews, but there is an error in one. When I added `async_` as a keyword in the function, tests hasn't been passed so it isn't functionally equivalent (commit "Just for a proof). 

And the second, there is an error with [pypy3](https://travis-ci.org/chtd/psycopg2cffi/builds/304745738?utm_source=github_status&utm_medium=notification) but  there aren't any errors with other python versions and even with pypy2.

unimariJo commented:
@lopuhin can you review changes?

unimariJo commented:
@lopuhin Could you please review my changes?

lopuhin commented:
@unimariJo thanks for fixing this issue! 👍
